### DISCUSSION TITLE
How to access full `pudl.sqlite` in a standard GitHub Runner

### DISCUSSION THREAD

**Discussion by bendnorman** [catalyst-cooperative/pudl#3071] [Ideas]
Since converting `PudlTabl` methods to dagster assets that write data to `pudl.sqlite`, `pudl.sqlite` has grown by about 10x. While having all of our data in a database rather than installing the `pudl` python package is more convenient for users, downloading an 11 GB database is cumbersome. This is especially a problem if you want to incorporate the full `pudl.sqlite` file into CI running on a standard Github Runner which only has [14 GB of disk space](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources).

There are a few future changes that should reduce the size of the database:
1. Removing `_out_*` assets from the database. These assets are intermediate steps to a more complete asset users should be interacting with. There are some `_out_*` in the database because they were available in `PudlTabl` and we want to continue to support `PudlTabl` until users migrate to pulling data directly from the database.
2. Converting some assets to SQL views. There are some assets that are simple joins that could be SQL views which would not increase the database's size.

Even if we reduce `pudl.sqlite`'s size by making these changes, there's no guarantee we keep the size down as we add more data sources. Here are a few ideas on how we can make this "medium" sized data easier to access:

1. Create a `pudl.sqlite` "lite" database which only includes `out_*` tables that include all of the most useful information.
2. Have users query `pudl.sqlite` using the [datasette API](https://docs.datasette.io/en/stable/json_api.html). Not the easiest method for pulling data into dataframes. We also don't have a great means of versioning the API.
3. Create a collection of parquet files for each table in `pudl.sqlite`. This would allow users to easily pull very space efficient files for individual tables into DataFrames. One potential issue with this solution is that users might end up pulling parquet files from different data versions.
4. Convert `pudl.sqlite` to a duckdb file. Given the design goals of duckdb it seems like an ideal tool for our type of data. However, I don't think duckdb's compression is as efficient as parquet's, the tooling only supports a database created with the same version and is an additional tool users will have to become familiar with. 
5. Providing a compressed version of `pudl.sqlite`. This will make it easier for folks to download but will still consume most of the GitHub runner disk space when unzipped.

**Comment by cmgosnell** [comment]
I'm not sure this is a good idea but i could imagine an additional idea being:

6. Create a `pudl.sqlite` "fast" database which only includes the few years we have in our fast settings file.

I personally think 1 or 6 but sounds the most ideal. 1 because that might be a nice way for most users to access our data regardless of github runners, but this could pose a problem if anyone wants to rely on the tidy normalized tables. and 6 because we already consistently build this db locally and for our CI so it sounds straightforward, but it would mean downstream users' tests would be locked to whatever years/partitions that we are currently testing.

**Reply by zaneselvans** [reply]
It seems like the options break down into a few categories:
- **Keep all the data in one file and make that file smaller by using compression or deduplication in the form of SQL views.**
  - Pro: maintains the relatively simple "download a file" method of distribution.
  - Pro: deduplication with views will probably speed up our ETL and provide less opportunity for ambiguity as to what the source of truth for a data point is.
  - Pro: Having all the data in a single file makes it easy to know it's all from the same version, and easy to cache on GitHub.
  - Con: allowing direct data access from a compressed file will either be a hack that slows down data access if we keep using SQLite, or will mean switching to another, less universally supported / familiar data format like DuckDB.
  - Con: it's not clear how much of our existing infrastructure will work smoothly with SQL Views.
- **Split the data into multiple files, such that users only need to download a subset of it for use in CI.**
  - Con: This would mean exploding the database into dozens or hundreds of Parquet files if we did it across the board.
  - Con: Alternatively, if we only break *some* of the big tables out into Parquet, then there's no uniform access method across the whole collection of data, we need another way to validate FK relationships, more people need to learn multiple methods of reading data.
  - Con: Greater potential for accidentally mixing-and-matching data from different versions.
  - Pro: A-la-carte data access means you're not downloading data you don't use.
  - Con: A-la-carte data access means you're probably downloading the same subset of data over and over again, unless we figure out a reliable caching mechanism. This might be slow enough to be annoying.
- **Allow remote access to data rather than copying it in bulk from one location to another.**
  - Con: SQLite is explicitly not designed to do this.  We would need another file format.
  - Pro: Parquet is very much designed to do this, and would enable lots of other remote accessibility applications.
  - Con: If this usage was popular, data egress might get expensive, unless it's coming out of AWS Open Data.
  - Con: Lots of data getting downloaded repeatedly, might be slow. Available RAM is also limited (even more so than disk).


Do we have a sense of how much space savings the various options would provide?
- `gzip pudl.sqlite` results in a file that is 2.7 GB, down from 12 GB. Takes 2-3 minutes to compress, ~10 seconds to extract. Pretty great! We should probably gzip all the SQLite DBs we distribute just so downloading is less painful. Unfortunately it doesn't sound like there's an easy way to read from SQLite without decompressing the whole file, so this doesn't address the CI problem.
- `xz --threads=8 pudl.sqlite` takes 9 minutes to compress, 1.5 minutes to decompress, and squeezes the file to 1.5GB.
- There's a [non-free SQLite extension](https://sqlite.org/com/cerod.html) that deals with compressed (and/or encrypted) SQLite DBs, and a not particularly maintained [open-source extension](https://phiresky.github.io/blog/2022/sqlite-zstd/) too. Both of those would kill the simplicity and universality of using SQLite to distribute the data.
- How much space could be saved with straightforward SQL Views?
- How small does the DB file need to be for it to actually be usable on a standard GitHub runner? 10GB? 7GB? Less?
- It seems like around ~10GB is the boundary where data is big enough that getting stuff to work for free becomes harder.
- Providing a straightforward way to access all of the data easily for use in CI or other applications seems like it's high value -- if we want people to build on the data in a programmatic way, this is kind of necessary.
- GitHub Actions are a pretty accessible platform for automation (without anybody needing to set up their own cloudy stuff).
- Storing the data in BigQuery is also pretty cheap, but someone would have to pay for the querying and egress.
- What's the real application that RMI is looking at right now, just testing? Or needing to actually work with all of the data?
- What's the relative cost of paying us to do the work to shrink the DB or create and maintain Parquet outputs, vs. paying for a larger GitHub runner?

It's not ideal but I suspect that paying for a larger runner, switching to distributing gzipped SQLite DBs, and caching the gzipped DB on GitHub to avoid downloading the same version over and over again would be a cheap, performant, and easy to implement solution.

**Reply by bendnorman** [reply]
I like the primary ideas y'all proposed:
1. Compress the sqlite files and use a large github runner if 14 GB isn't enough. Hopefully, our table trimming and SQL view conversion will bring the size back down.
2. Offer a pudl.sqlite "Lite" version that contains the most useful `out_` tables. I think this would serve users who are only interested in the more denormalized and complete tables. I do think a PUDL lite version would require some thought as to what tables should be included and how to describe the distinction between the full and lite versions. I'm a little wary about releasing `pudl.sqlite` produced with the `etl_fast.yml` because some users might want different subsets of years.

I think option 1 is the most straightforward in the short term. Long term I think we'll need to do more user research to understand the best modes of access.

**Reply by zaneselvans** [reply]
I wrote a quickie script to convert our SQLite DB to DuckDB. It takes 2 minutes to run and the DuckDB file is only 2.6 GB.  Note that it doesn't preserve the extensive constraints that we have defined in the SQLite DB, but definitely something we could look at in the future...

```py
import click
import logging
import duckdb
from pathlib import Path

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


@click.command()
@click.argument("sqlite_path", type=click.Path(exists=True, resolve_path=True))
@click.argument(
    "duckdb_path", type=click.Path(resolve_path=True, writable=True, allow_dash=False)
)
def convert_sqlite_to_duckdb(sqlite_path, duckdb_path):
    """
    Convert an SQLite database to DuckDB format.

    Args:
        sqlite_path (str): Path to the existing SQLite database file.
        duckdb_path (str): Path to the new DuckDB database file (should not exist).

    Example:
        python sqlite_to_duckdb.py sqlite.db duckdb.db
    """
    sqlite_path = Path(sqlite_path)
    duckdb_path = Path(duckdb_path)

    # Check if DuckDB file already exists
    if duckdb_path.exists():
        click.echo(
            f"Error: DuckDB file '{duckdb_path}' already exists. Please provide a new filename."
        )
        return

    # Connect to DuckDB database
    duckdb_conn = duckdb.connect(database=str(duckdb_path))
    duckdb_cursor = duckdb_conn.cursor()

    # Fetch table names from SQLite database using DuckDB
    duckdb_cursor.execute(f"ATTACH DATABASE '{sqlite_path}' AS sqlite_db;")
    duckdb_cursor.execute("SELECT name FROM main.sqlite_master WHERE type='table';")
    table_names = [row[0] for row in duckdb_cursor.fetchall()]

    # Copy tables from SQLite to DuckDB
    for table_name in table_names:
        logger.info(f"Working on table: {table_name}")
        # Fetch column names and types from SQLite table using DuckDB
        duckdb_cursor.execute(f"PRAGMA table_info(sqlite_db.{table_name});")
        columns_info = duckdb_cursor.fetchall()
        column_definitions = ", ".join([f"{col[1]} {col[2]}" for col in columns_info])

        # Create equivalent table in DuckDB
        duckdb_cursor.execute(f"CREATE TABLE {table_name} ({column_definitions});")

        # Copy data from SQLite to DuckDB using DuckDB
        duckdb_cursor.execute(
            f"INSERT INTO {table_name} SELECT * FROM sqlite_db.{table_name};"
        )

    # Commit and close connections
    duckdb_conn.commit()
    duckdb_conn.close()


if __name__ == "__main__":
    convert_sqlite_to_duckdb()
```

**Reply by arengel** [reply]
I wanted to offer an alternative proposal that I shared with @cmgosnell and @bendnorman during our call today:

7. Create a zip archive of parquets.

- Pro: Zip archives and parquet files are established and stable standards with wide support and basic use requires no additional dependencies for users (or PUDL).
- Pro: Compact, last time I tested it, the resulting zip file was 1.5 GB built from a 12+GB sqlite.
- Pro: As a single file, it is easy to download the whole thing if you like, and easy to ensure that all the contained tables are consistent, i.e. built from the same version of PUDL, and easily versioned.
- Pro: It can be stored and distributed just like the sqlite.
- Pro: Remote zip files can be opened programmatically and individual files can be extracted without downloading the whole zip archive. Offering some of the benefits of a hosted database. See https://github.com/gtsystem/python-remotezip. There are some server-side requirements but from the examples it appears that it works on S3.
- Con: This is not a standard way of distributing a collection of tables / database.
- Con: It's not a database, and there is no existing way (as far as I am aware) to interact with it as if it were one. 


I haven't investigated the memory requirements or other performance characteristics of this solution but we use a similar setup for serializing python objects and haven't run into memory issues.

It's pretty straightforward to create this from the sqlite but it would be preferable to create the parquets directly to preserve richer dtypes.

``` python
import sqlite3
import pandas as pd

from io import BytesIO
from pathlib import Path
from zipfile import ZipFile

# Write all tables to zip
tables = sqlite3.connect(sql_path).cursor().execute(
    "SELECT name FROM sqlite_master WHERE type='table';"
).fetchall()
with ZipFile(zip_path, mode="w") as z:
    for (table,) in tables:
        z.writestr(
            f"{table}.parquet",
            pd.read_sql_table(table, f"sqlite:///{sql_path}").to_parquet(),
        )


# Read a table from zip
with ZipFile(zip_path) as z:
    df = pd.read_parquet(BytesIO(z.read(f"{table}.parquet")))
```

**Reply by zaneselvans** [reply]
* Parquet files are already compressed internally by default so the primary benefit of zipping a bunch of them together would be to bundle them into a single unit, rather than to make the filesize smaller.
* Parquet files are specifically designed for efficient remote operations with minimal data transfer, so if we were to go to the trouble of exporting all of the data in Parquet, it seems like we ought to provide direct access to the Parquet for easy native access through all the tools that understand the format, without requiring anybody to deal with the zipfile. @simonw recently wrote up [a nice little series of examples](https://til.simonwillison.net/duckdb/remote-parquet) using DuckDB, polars, and other tools to do this. Pandas and Dask are also happy to do it.
* Extracting whole Parquet files from the zip archive would be much less efficient than allowing people to run queries that only depend on a subset of the columns and row-groups in the Parquet files.
* Organizationally, many of the benefits of zipping the parquet files together would also come from having them in a single directory together, and that whole directory can be downloaded recursively with the `gcloud storage` or `aws s3` CLIs. I guess we could also provide a one-file archive of the exact same files for download over `https://`.
* The much richer PyArrow datatypes that we would have access to in Parquet would be great, and will line up exactly with Pandas' move toward PyArrow as first-class backend to dataframes (thought it's still experimental right now).
* DuckDB (as a query engine) is perfectly happy to treat a directory of parquet files (local or remote) like a database, but as mentioned above we'll lose all of the inter-table relational information and many of the `CHECK` constraints that are currently embedded in the SQLite DB (though the `ENUM` constraints would translate into very efficient categorical columns in Parquet).
* A directory of Parquet files is a very common way to publish tabular data for cloud-native use, and would probably enable lots of good use cases. We can distribute it for free on the AWS Open Data Registry and not have to worry about data egress costs. Maybe we should do it and just leave it up to the users to manage different versions if they actually download the files.
* To Archive directories of hundreds of files on Zenodo we'll need to bundle them into zipfiles / tarballs.
* Structurally we have the infrastructure in place to do it (a Parquet IOManager, and a method to generate PyArrow schemas based on our table-levelmetadata) but I know it doesn't just work out of the box right now. The `Resource.to_pyarrow()` schema generation method or the metadata it consumes would need some work to apply generally to all of the assets in the DB (currently we only use it on the EPA CEMS).

**Reply by gschivley** [reply]
@zaneselvans I'm all for using duckdb storage but is it mature enough to depend on yet? The one thing you *can't* find in their documentation is info about duckdb file storage. Other than the [warning](https://duckdb.org/internals/storage) that it will be in flux until 1.0.0.

**Reply by jdangerx** [reply]
I think the idea with DuckDB here would be to use it just as a frontend for querying a bunch of different Parquet files - so the storage would be in a nice stable format. And if DuckDB implodes for some reason there are other tools to access parquet files.

If we wanted to maintain some set of SQL constraints on top of the pile of Parquet files, then I think we would need to distribute a duckdb file... which then does run into your stability concerns. But we could just not worry ourselves with SQL constraints for now.

**Reply by zaneselvans** [reply]
I think we definitely need to keep worrying about the SQL constraints on the data **production** side, even if they're not encoded into all of the data products which we distribute, since they help us catch many, many internal data inconsistencies and bugs. But if we keep producing a unified `pudl.sqlite(.gz)` output alongside a directory of Parquet files derived from the same data, then we would know that those internal constraints had been satisfied.

**Reply by jdangerx** [reply]
Yeah, totally agreed. In theory we could maintain these constraints while moving over to `pudl.duckdb` but after a whole bunch of research it still feels like a lot of effort for questionable gain on the production side. But Parquet files on the distribution side seems like a no-brainer at this point.

We could potentially produce a constraints+schema only version of the duckdb that downstream users could then import data into - not really sure what that buys the users, though - and definitely is a weird layer of complexity. So I'm inclined against that.

**Reply by zaneselvans** [reply]
I don't think it would be *too* much work to output a DuckDB schema. We should replace all of our `to_sql()` methods with `to_sqlite()` and create `to_duckdb()` methods that translate the types and constraints correctly for the DuckDB dialect of Postgres.  At which point we would basically have the ingredients for a `DuckDBIOManager` if we wanted to implement it.

Many of these tasks sounds like they could be a good fit for a data-warehouse / SQL savvy open source contributor if say Brent or Jan or someone like that were interested.

**Comment by gschivley** [comment]
@bendnorman you mention in 3) 
> One potential issue with this solution is that users might end up pulling parquet files from different data versions.

How are data versions tracked right now? When I look at the nightly build or Datasette tables there doesn't seem to be an identifier with data versioning. Is it just that a complete `pudl.sqlite` database is going to be internally consistent, and individual tables downloaded at different times might not be?

I have users that access a handful of PUDL tables for things like unit definitions, capacity, heat rates, historical capacity factor, etc. I'm only using the annual EIA data and supplement it with 860m from my own processing pipeline. At the moment I ask users to download PUDL from Zenodo and they probably use the same database until a change in table/column names prompts me to tell everyone to download a new version (this may have only happened once...).

My ideal workflow would be fetching data as users need it to minimize the up-front data compilation needs. Each time a user runs my software it would run a quick query (count of rows or max `report_date`) to see if the cloud version is updated from the local. Looks like it's easy to run `SELECT COUNT(*) FROM <file.parquet>` with duckdb, which uses the file metadata rather than downloading any of the data. So individual parquet files would be great from my perspective. A zip file of individual parquet files could also work if there were a metadata file describing when each was last modified.

**Reply by jdangerx** [reply]
> How are data versions tracked right now? When I look at the nightly build or Datasette tables there doesn't seem to be an identifier with data versioning. Is it just that a complete `pudl.sqlite` database is going to be internally consistent, and individual tables downloaded at different times might not be?

Yeah, exactly right. Data versions aren't really tracked but we rely on the bundling that `pudl.sqlite` provides to make sure everything is internally ocnsistent.

> My ideal workflow would be fetching data as users need it to minimize the up-front data compilation needs. Each time a user runs my software it would run a quick query (count of rows or max `report_date`) to see if the cloud version is updated from the local. Looks like it's easy to run `SELECT COUNT(*) FROM <file.parquet>` with duckdb, which uses the file metadata rather than downloading any of the data. So individual parquet files would be great from my perspective. A zip file of individual parquet files could also work if there were a metadata file describing when each was last modified.

If downloading multiple tables, you should probably redownload *all* of them every time *any* of them appears out of date, to make sure that you're getting internally consistent data. But otherwise this sounds reasonable! You might want to use modified-time or file-hash instead of a query to determine staleness.

**Reply by bendnorman** [reply]
We haven't done regular data releases of PUDL data (about once a year at this point) which is why datasette and our documentation refer to data produced by the `dev` branch. We haven't done more frequent releases because packaging the software and data and releasing on Zenodo is burdensome, which is part of the reason for moving to data only releases. Our goal with data only releases is to have bi weekly or monthly data releases with version numbers. This way, users can pin to specific data versions and won't get sabotaged by any breaking changes on `dev`.

**Comment by jdangerx** [comment]
It sounds like there's appetite for distributing a pile of parquet files, while punting any decisions about DuckDB usage on the Catalyst side - people are obviously free to use whatever tooling fits the bill (ha) for accessing data in parquet files.

It also sounds like we should continue distributing the full SQLite file for people who want SQL constraints, and certainly continue using it internally so we can use those constraints while writing the dang data.

I'm hearing a little less support for the "pre-subsetted" SQLite files (`out_*`, `etl_fast`) and for compressing the existing SQLite files, though I know compressing the existing files will at least make it easier for folks like @zaneselvans and @aesharpe who often have sketchy internet connections to download the DB.

I sort of think all of these are pretty easy to generate from our existing uncompressed complete SQLite output:
* a pile of parquet files
* compressed SQLite files
* a few pre-subsetted SQLite files

And the hard thing is "figure out if we want to move pudl.sqlite completely over to duckdb."

So my proposal is to do the three easy things but not the hard thing. 

In terms of ordering I think we should do the Parquet stuff first since that seems to solve more downstream problems than the other two.

**Reply by zaneselvans** [reply]
* 👍🏼 to compressed SQLite. I think we should probably do this first since it ought to be a one-liner in the nightly build script and reduce download time/size by like 75% without materially affecting data accessibility.
* 👍🏼 to directory-of-parquet-files output alongside SQLite, since this opens up a lot of very convenient a-la-carte data usage and cloud-native analysis, and costs us next to nothing in storage & compute. I suspect it will be a bit of work to get the PyArrow schema generation working and to do double outputs during the ETL (with both SQLite and Parquet IOManagers). Also note that without significant additional work, the data validations will continue to run only on the SQLite outputs, and if there are issues specific to the Parquet outputs we won't know about them. But if we get the NSF resources to invest in the data validation refactor we'll hopefully be able to address this next year?
* 👎🏼 to the pre-subsetted SQLite files. It seems like substantial added potential complexity / confusion around where data comes from and what data you can assume is available, and I **think** many of the benefits this would provide are available from simply compressing the SQLite database, or allowing individual table access through Parquet.

If we get the Parquet outputs working, we and users will have lots of opportunity to play with DuckDB querying them either locally or remotely, and if/when their storage format calms down all of that experience should translate directly to using a unified DuckDB output that could potentially replace the SQLite output.

Another thing that would help all of these options be more legible is publishing structured metadata alongside the main PUDL data outputs. Upgrading to frictionless v5 would allow us to publish datapackage.json files with all the constraints and table/column descriptions annotating both the SQLite DB and a parquet directory.

**Reply by jdangerx** [reply]
Yeah, pre-subsetted SQLite files definitely seems like the iffiest one. I was just thinking about how we'd have to rewrite the data access documentation and that seemed like a headache. Down to just do parquet + compression.

**Reply by bendnorman** [reply]
I'm in favor of distributing a compressed SQLite database and a directory of parquet files! My only concern is having the .sqlite db and the parquet files will confuse users. This can probably be resolved by having clear docs: "If you want to do X,Y,Z use .sqlite db. If you want to do A, B, C, use parquet files.".

**Comment by bendnorman** [comment]
I ran `sqlite3_analyzer`on the most recent full `pudl.sqlite` and it looks like the two space saving measures I mentioned in the original comment won't make a significant dent in the size of the db. Most of the largest tables are not intermediate tables or good candidates for SQL views. 

Intermediate output tables (denoted by `_out_*`) that we plan to remove account for 9.2% of the database. The 8 largest tables account for 60% of the database. None of these are great candidates for SQL views.

Sounds like we've settled on compressing the SQLite db and distributing collections of Parquet files but I thought this was mildly helpful information. I've attached the full disk utilization report: 
[pudl_sqlite_analyze.txt](https://github.com/catalyst-cooperative/pudl/files/13443746/pudl_sqlite_analyze.txt)

**Reply by zaneselvans** [reply]
Thanks for checking!

I'm pretty happy with our action items.
